instance/models/server.py

```diff
@@ -427,17 +427,25 @@ def reboot(self, reboot_type='SOFT'):
 
     def terminate(self):
         """
-        Terminate the server
+        Stop and terminate the server.
+
+        We explicitly stop and wait for a graceful shutdown of the server before terminating it.
+        This ensures any daemons that need to perform cleanup tasks can do so, or that any
+        shutdown scripts/tasks get executed.
         """
-        self.logger.info('Terminating server (status=%s)...', self.status)
         if self.status == Status.Terminated:
             return
-        elif not self.vm_created:
+
+        self.logger.info('Terminating server (status=%s)...', self.status)
+        if not self.vm_created:
+            self.logger.info('Note: server was not created when terminated.')
             self._status_to_terminated()
             return
 
         try:
-            self.os_server.delete()
+            os_server = self.os_server
+            self._shutdown(os_server)
```

Comment:
We can do a small check here to see if it wasn't shut down, so we can log a message notifying the user:

```python
if not self._shutdown(os_server):
    self.logger.warning("Server has not reached SHUTOFF state after max wait time; terminating forcefully.")
```